event-store/src/adapters/amqp/emitter.rs

```diff
@@ -0,0 +1,182 @@
+                }).and_then(move |channel: Channel<TcpStream>| {
+                    let ch = channel.clone();
+                    tx.send(ch).expect("Send channel to main thread");
+                    channel.exchange_declare(
+                        &exchange,
+                        &"topic",
+                        ExchangeDeclareOptions {
+                            durable: true,
+                            ..ExchangeDeclareOptions::default()
+                        },
+                        FieldTable::new(),
+                    )
+                }).map_err(|e| {
+                    error!("Error connecting to AMQP: {}", e);
+                }),
+        );
+        let channel = rx
+            .recv_timeout(Duration::from_secs(5))
+            .map_err(|e| {
+                error!("AMQP channel was not made available: {}", e);
+                panic!(e);
+            }).expect("AMQP channel was not made available");
+        Self {
+            channel,
+            namespace,
+            exchange: _exchange,
+        }
+    }
+}
+
+fn prepare_subscription<E, H>(
+    queue_name: String,
+    event_name: String,
+    exchange: String,
+    handler: H,
+    channel: Channel<TcpStream>,
+) -> impl Future<Item = (), Error = io::Error>
+where
+    E: EventData,
+    H: Fn(&Event<E>) -> (),
+{
+    let c_channel = channel.clone();
+    let queue_name1 = queue_name.clone();
+    info!("Creating queue {}", queue_name);
+    channel
+        .queue_declare(
+            &queue_name,
+            QueueDeclareOptions {
+                durable: true,
+                exclusive: false,
+                auto_delete: false,
+                ..QueueDeclareOptions::default()
+            },
+            FieldTable::new(),
+        ).and_then(move |queue| {
+            info!("Binding queue {} to exchange {}", queue_name, exchange);
+            channel
+                .queue_bind(
+                    &queue_name,
+                    &exchange,
+                    &event_name,
+                    QueueBindOptions::default(),
+                    FieldTable::new(),
+                ).and_then(move |_| {
+                    channel.basic_consume(
+                        &queue,
+                        &queue_name,
+                        BasicConsumeOptions::default(),
+                        FieldTable::new(),
+                    )
+                }).and_then(move |stream| {
+                    info!("Starting to consume from queue {}", queue_name1);
+                    stream.for_each(move |message| {
+                        let data: Event<E> =
+                            serde_json::from_str(str::from_utf8(&message.data).unwrap()).unwrap();
+                        info!("Receiving message with id {}", data.id);
+                        handler(&data);
+                        c_channel.basic_ack(message.delivery_tag, false)
+                    })
+                })
+        }).and_then(|_| ok(()))
```

Comment:
`ok(())` is confusing. Me and @petereast found examples that import it as `use futures::future::{ok as FutOk, Future};`, that way you can do `FutOk(())` which looks a bit more like a `Result`, which is kinda what it is.

Reply:
dONE